### ISSUE TITLE
[QUESTION] Non-convergence issue when training with --bf16 enabled

### ISSUE THREAD

**Issue by Centaurus982034** [thu-ml/ReMoE#11]
Hi Ziteng,

I encountered a convergence issue when running scripts/train_llama_182m_moe.sh.

When running with the default setting (where --bf16 is enabled), the model fails to converge, and the LM loss remains stuck above 5, and the grad norm explodes. However, if I disable --bf16, the model converges as expected.

I observed the same behavior when running scripts/train_llama_182m_remoe.sh.

Environment:

Docker Image: nvcr.io/nvidia/pytorch:24.08-py3

Do you have any insights into what might be causing this instability with BF16?

Thanks!

<img width="1922" height="1002" alt="Image" src="https://github.com/user-attachments/assets/53f036b0-81d6-4577-a6c8-1cc569b5e5e4" />

<img width="1980" height="1056" alt="Image" src="https://github.com/user-attachments/assets/b02f899e-c43f-4a65-89c3-5136202e6910" />

**Comment by thuwzt** [comment]
Issue #9 also reported divergence. You can follow the software environment specified in [this comment](https://github.com/thu-ml/ReMoE/issues/9#issuecomment-3326323971).

For the precision issue, this might be related to specific implementation of GroupedGEMM on certain hardware. Could you provide the model of GPU you are using?

**Comment by Centaurus982034** [comment]
Thanks.

But most of the environments are identical to yours, including transformer_engine, torch, flash-attn, apex, and grouped-gemm. In fact, I tried to train with --bf16 on and --moe-grouped-gemm off, the training also diverges. My GPU is NVIDIA H800 with driver version 535.161.08.

**Comment by thuwzt** [comment]
This seems weird. The experiments in the paper use NVIDIA A100 and the loss converges. But based on your trials, the MoE baseline fails when using bf16, even when not using GroupedGEMM. I suspect a compatibility issue of the specific commit of Megatron on Hopper, given that Megatron’s master branch is updated very frequently.

A potential fix would be to rebase Megatron onto a recent master commit and verify the convergence of vanilla MoE model with [the script](https://github.com/thu-ml/ReMoE/blob/main/scripts/train_llama_182m_moe.sh) in this repo. If the MoE baseline converges, then you can cherry-pick [this commit](https://github.com/thu-ml/ReMoE/commit/c9d3f7940f744e2cdd90e7fea6dea43b793ff6ca) (directly copy-paste the added lines) to support ReMoE. The modifications are plug-and-play with minimal changes: `megatron/core/transformer/moe/router.py` to add ReLU routing and `megatron/training/training.py` to apply L1 regularization loss, with other sanity adjustments such as adding configs.
